### ISSUE TITLE
JAX-aware Pickler / Unpickler Code Review

### ISSUE THREAD

**Issue by JossWhittle** [JossWhittle/Opt-ID#2]
Tensors in JAX are synonymous to Numpy arrays but have the additional property of living on a physical device such as cpu:0 or gpu:0.

Currently, when a JAX tensor is pickled it decomposes into a native Numpy array which means it is no longer a JAX DeviceArray when it is unpickled and the device it originally lived on has been lost.

```python
tensor = jnp.eye(4) # 4x4 Identity Matrix

print('before', type(tensor), str(tensor.device_buffer.device()), tensor.dtype)

with open('tensor.pkl', 'wb') as fp:
    pickle.dump(tensor, fp)
    
with open('tensor.pkl', 'rb') as fp:
    tensor = pickle.load(fp)
    
print('after', type(tensor), tensor.dtype)
```

```
before <class 'jax.interpreters.xla._DeviceArray'> cpu:0 float32
after  <class 'numpy.ndarray'> float64
```

Using JAX-aware inherited pickler classes, during pickling, when a JAX DeviceArray is encountered it is replaced by a tuple.

```(tag: str, (device_name: str, values: np.ndarray))``` 

When tuples with this structure are encountered during unpickling, and the tag matches the expected value, the tuple is merged back together and a JAX DeviceArray on the correct device is returned. 

```python
tensor = jnp.eye(4) # 4x4 Identity Matrix

print('before', type(tensor), str(tensor.device_buffer.device()), tensor.dtype)

with open('tensor.pkl', 'wb') as fp:
    optid.utils.io.JAXPickler(fp).dump(tensor)
    
with open('tensor.pkl', 'rb') as fp:
    tensor = optid.utils.io.JAXUnpickler(fp).load()
    
print('after', type(tensor), str(tensor.device_buffer.device()), tensor.dtype)
```

```
before <class 'jax.interpreters.xla._DeviceArray'> cpu:0 float32
after  <class 'jax.interpreters.xla._DeviceArray'> cpu:0 float32
```

---

Code:

https://github.com/JossWhittle/Opt-ID/blob/8605f28473ec5ee2e4231d1dae3403f7e32a1f33/src/optid/utils/io.py#L26-L254

Test suite:

https://github.com/JossWhittle/Opt-ID/blob/8605f28473ec5ee2e4231d1dae3403f7e32a1f33/tests/utils/test_io.py#L34-L303

---

Test suite maxes out at 97% coverage because we can't test that an exception is thrown when reading in a ShardedDeviceArray (which is not supported currently) unless it was possible to write a ShardedDeviceArray before hand (which is also not supported currently).

**Comment by LAShemilt** [comment]
Hi Joss,

This all makes sense to me,  - are you planning to put this into the Jax package at all? It could be useful I think for many people. I have copied your JaxPickler and JaxUnipickler class locally and run some of your examples. All works for me. 

I don't have many comments on the code itself it all looks fine to me.

**Comment by JossWhittle** [comment]
Thanks, Laura.

That's the plan. There are other ways of injecting custom behaviour into pickle that I experimented with but I think this is the best solution for JAX as it allows the opportunity to inject a mapping for device names while none of the other approaches I tried allowed this.

I'm not sure if the ShardedDeviceArray handler is robust enough, I was digging through the JAX source and there are a lot of different cases the ShardedDeviceArray can handle for how data is split across devices, but I'm not sure if these actually occur in practice for variables that end up in user land. 

It seems like if you've managed to get a ShardedDeviceArray to be returned to you (even if there was more exotic sharding during the computation that generated it) it will always be sharded trivially across the first axis where the number of devices is the first axis size, so for now I think it will work fine.

**Comment by LAShemilt** [comment]
I think that the tuple is nice as the biggest part of the data object is the array. A tuple is lightweight and easy to manipulate. 

I can test ShardedDeviceArray a bit more if you want . I only used your code examples in the first part of this issue locally and CPU only. I guess the admins on the Jax repo can help guide you as to what they would expect from the code.
